File: lib/src/choose_pseudo_count.cpp
#include <vector>

#include "pybind11/pybind11.h"
#include "pybind11/numpy.h"
#include "scran_norm/scran_norm.hpp"

#include "utils.h"

double choose_pseudo_count(const pybind11::array& size_factors, double quantile, double max_bias, double min_value) {
    scran_norm::ChoosePseudoCountOptions opt;
    opt.quantile = quantile;
    opt.max_bias = max_bias;
    opt.min_value = min_value;
    return scran_norm::choose_pseudo_count(size_factors.size(), check_numpy_array<double>(size_factors), opt);
}

void init_choose_pseudo_count(pybind11::module& m) {
    m.def("choose_pseudo_count", &choose_pseudo_count);
}

File: src/scranpy/choose_pseudo_count.py
import numpy

from . import lib_scranpy as lib


def choose_pseudo_count(
    size_factors: numpy.ndarray,
    quantile: float = 0.05, 
    max_bias: float = 1,
    min_value: float = 1
) -> float:
    """Choose a suitable pseudo-count to control the bias introduced by
    log-transformation of normalized counts.

    Args:
        size_factors:
            Array of size factors for all cells.

        quantile:
            Quantile to use for defining extreme size factors.

        max_bias:
            Maximum allowed bias in the log-fold changes between cells.

        min_value: 
            Minimum value for the pseudo-count.

    Returns:
        Choice of pseudo-count, for use in
        :py:func:`~scranpy.normalize_counts.normalize_counts`.
    """
    local_sf = numpy.array(size_factors, dtype=numpy.float64, copy=False)
    return lib.choose_pseudo_count(local_sf, quantile, max_bias, min_value)
